Reindex search resiliency

henningandersen opened the pull request:
Local reindex can now survive loosing data nodes that contain source
data. The original query will be restarted with a filter for
`_seq_no >= last_seq_no` when a failure is detected.

Part of #42612 and split out from #43187

A couple of follow-ups will be done (all indicated with todos) in addition to what is in the meta issue:

1. Fix the problem with `allowPartialSearchResult` in master
2. Build the `ScrollableHitSource` separately rather than in super constructor.
3. SeqNo mapping creation or avoid the issue.

henningandersen committed:
Reindex search resiliency

Local reindex can now survive loosing data nodes that contain source
data. The original query will be restarted with a filter for
`_seq_no >= last_seq_no` when a failure is detected.

Part of #42612 and split out from #43187

elasticmachine commented:
Pinging @elastic/es-distributed

tbrooks8 commented:
I will take a look at this tomorrow.

tbrooks8 commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkByScrollAction.java:
```diff
@@ -135,7 +136,9 @@
         BackoffPolicy backoffPolicy = buildBackoffPolicy();
         bulkRetry = new Retry(BackoffPolicy.wrap(backoffPolicy, worker::countBulkRetry), threadPool);
-        scrollSource = buildScrollableResultSource(backoffPolicy);
+        // todo: this is trappy, since if a subclass override relies on subclass fields, they are not initialized. We should fix
```
I agree. This was causing me problems in a recent PR.

Tim-Brooks reviewed (commented):
This is mostly looking good to me. I don't see any issues with the approach or your design. It seems likely that the test failures are related to this change. Do you want to let me know when I should look again to officially approve?

henningandersen committed:
Merge remote-tracking branch 'origin/reindex_v2' into enhance_reindex_search_resiliency

henningandersen committed:
Reindex search resiliency

No longer retry/restart the original search request, since this is not
what we used to do and it leads to long wait time to get the info back
that a search request is bad.

henningandersen commented:
Thanks @tbrooks8 and sorry for not fixing the test failure before now. It should be ready for another round (provided tests succeed).

henningandersen committed:
Reindex search resiliency

No longer fail on the empty index. So far we consider this a
workaround/hack more than the solution.

henningandersen committed:
Reindex search resiliency

Use unmapped_type to circumvent problem with newly created indices
without a mapping, since this is local to reindex. Once type removal is
complete, ensuring that the metadata fields are always available likely
becomes easier, so we will defer a solution to the search/sort problem.

henningandersen commented:
@elasticmachine run elasticsearch-ci/1

Tim-Brooks reviewed (approved):
LGTM
